pyro/ops/contract.py

```diff
@@ -56,12 +56,14 @@ class TensorRing(object):
     def __init__(self, cache=None):
         self._cache = {} if cache is None else cache
 
-    def _save_tensor(self, tensor):
+    def _hash_by_id(self, tensor):
         """
-        Saves a tensor in the cache so that ``id(tensor)`` can be used as a
-        key in the cache without risk if the id being recycled.
+        Returns the id of a tensor and saves the tensor so that this id can be
+        used as a key in the cache without risk of the id being recycled.
         """
-        self._cache['tensor', id(tensor)] = tensor
+        result = id(tensor)
+        assert self._cache.setdefault(('tensor', result), tensor) is tensor
```

Comment:
Is the cache bounded in any way, since we might be holding tensors in memory now that might have been previously GCd?

Reply:
The cache is not bounded. This is the einsum sharing cache whose lifetime is the duration of a single `ubersum()` call (or `contract_to_tensor()` call or `svi.step()`). That lifetime is required anyway for `.backward()`; while some CPU tensors may have been deleted, their underlying storage would still remain until after `.backward()`.

Also, any previous recycling was incorrect. I believe it is only the `modulo_total` trick in this PR that allowed recycling, and that is why I've included both bugfix + feature in this PR.